ath79: Add support for OpenMesh MR1750 family

ecsv opened the pull request:
All these devices were part of the MR1750 image in the ar71xx target. These were now converted to ath79 and split into different images in this process

OpenMesh MR1750 v1
==================

Device specifications:
----------------------

* Qualcomm/Atheros QCA9558 ver 1 rev 0
* 720/600/240 MHz (CPU/DDR/AHB)
* 128 MB of RAM
* 16 MB of SPI NOR flash
  - 2x 7 MB available; but one of the 7 MB regions is the recovery image
* 3T3R 2.4 GHz Wi-Fi (11n)
* 3T3R 5 GHz Wi-Fi (11ac)
* 6x GPIO-LEDs (2x wifi, 2x status, 1x lan, 1x power)
* 1x GPIO-button (reset)
* external h/w watchdog (enabled by default))
* TTL pins are on board (arrow points to VCC, then follows: GND, TX, RX)
* 1x ethernet
    + 10/100/1000 Mbps Ethernet
    + 802.3af POE
    + used as LAN interface
* 12-24V 1A DC
* internal antennas

Flashing instructions:
----------------------

Various methods can be used to install the actual image on the flash.
Two easy ones are:

### ap51-flash

The tool ap51-flash (https://github.com/ap51-flash/ap51-flash) should be used to transfer the image to the u-boot when the device boots up.


### initramfs from TFTP

The serial console must be used to access the u-boot shell during bootup. It can then be used to first boot up the initramfs image from a TFTP server (here with the IP 192.168.1.21):

    setenv serverip 192.168.1.21
    setenv ipaddr 192.168.1.1
    tftpboot 0c00000 <filename-of-initramfs-kernel>.bin && bootm $fileaddr

The actual sysupgrade image can then be transferred (on the LAN port) to the device via

    scp <filename-of-squashfs-sysupgrade>.bin root@192.168.1.1:/tmp/

On the device, the sysupgrade must then be started using

    sysupgrade -n /tmp/<filename-of-squashfs-sysupgrade>.bin


OpenMesh MR1750 v2
==================

Device specifications:
----------------------

* Qualcomm/Atheros QCA9558 ver 1 rev 0
* 720/600/240 MHz (CPU/DDR/AHB)
* 128 MB of RAM
* 16 MB of SPI NOR flash
  - 2x 7 MB available; but one of the 7 MB regions is the recovery image
* 3T3R 2.4 GHz Wi-Fi (11n)
* 3T3R 5 GHz Wi-Fi (11ac)
* 6x GPIO-LEDs (2x wifi, 2x status, 1x lan, 1x power)
* 1x GPIO-button (reset)
* external h/w watchdog (enabled by default))
* TTL pins are on board (arrow points to VCC, then follows: GND, TX, RX)
* 1x ethernet
    + 10/100/1000 Mbps Ethernet
    + 802.3af POE
    + used as LAN interface
* 12-24V 1A DC
* internal antennas

Flashing instructions:
----------------------

Various methods can be used to install the actual image on the flash.
Two easy ones are:

### ap51-flash

The tool ap51-flash (https://github.com/ap51-flash/ap51-flash) should be used to transfer the image to the u-boot when the device boots up.


### initramfs from TFTP

The serial console must be used to access the u-boot shell during bootup. It can then be used to first boot up the initramfs image from a TFTP server (here with the IP 192.168.1.21):

    setenv serverip 192.168.1.21
    setenv ipaddr 192.168.1.1
    tftpboot 0c00000 <filename-of-initramfs-kernel>.bin && bootm $fileaddr

The actual sysupgrade image can then be transferred (on the LAN port) to the device via

    scp <filename-of-squashfs-sysupgrade>.bin root@192.168.1.1:/tmp/

On the device, the sysupgrade must then be started using

    sysupgrade -n /tmp/<filename-of-squashfs-sysupgrade>.bin

ecsv committed:
ath79: Add support for OpenMesh MR1750 v2

Device specifications:
======================

* Qualcomm/Atheros QCA9558 ver 1 rev 0
* 720/600/240 MHz (CPU/DDR/AHB)
* 128 MB of RAM
* 16 MB of SPI NOR flash
  - 2x 7 MB available; but one of the 7 MB regions is the recovery image
* 3T3R 2.4 GHz Wi-Fi (11n)
* 3T3R 5 GHz Wi-Fi (11ac)
* 6x GPIO-LEDs (2x wifi, 2x status, 1x lan, 1x power)
* 1x GPIO-button (reset)
* external h/w watchdog (enabled by default))
* TTL pins are on board (arrow points to VCC, then follows: GND, TX, RX)
* 1x ethernet
  - AR8035 ethernet PHY (RGMII)
  - 10/100/1000 Mbps Ethernet
  - 802.3af POE
  - used as LAN interface
* 12-24V 1A DC
* internal antennas

Flashing instructions:
======================

Various methods can be used to install the actual image on the flash.
Two easy ones are:

ap51-flash
----------

The tool ap51-flash (https://github.com/ap51-flash/ap51-flash) should be
used to transfer the image to the u-boot when the device boots up.

initramfs from TFTP
-------------------

The serial console must be used to access the u-boot shell during bootup.
It can then be used to first boot up the initramfs image from a TFTP server
(here with the IP 192.168.1.21):

   setenv serverip 192.168.1.21
   setenv ipaddr 192.168.1.1
   tftpboot 0c00000 <filename-of-initramfs-kernel>.bin && bootm $fileaddr

The actual sysupgrade image can then be transferred (on the LAN port) to the
device via

  scp <filename-of-squashfs-sysupgrade>.bin root@192.168.1.1:/tmp/

On the device, the sysupgrade must then be started using

  sysupgrade -n /tmp/<filename-of-squashfs-sysupgrade>.bin

Signed-off-by: Sven Eckelmann <sven@narfation.org>
[rebase, add LED migration]
Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

ecsv committed:
ath79: Add support for OpenMesh MR1750 v1

Device specifications:
======================

* Qualcomm/Atheros QCA9558 ver 1 rev 0
* 720/600/240 MHz (CPU/DDR/AHB)
* 128 MB of RAM
* 16 MB of SPI NOR flash
  - 2x 7 MB available; but one of the 7 MB regions is the recovery image
* 3T3R 2.4 GHz Wi-Fi (11n)
* 3T3R 5 GHz Wi-Fi (11ac)
* 6x GPIO-LEDs (2x wifi, 2x status, 1x lan, 1x power)
* 1x GPIO-button (reset)
* external h/w watchdog (enabled by default))
* TTL pins are on board (arrow points to VCC, then follows: GND, TX, RX)
* 1x ethernet
  - AR8035 ethernet PHY (RGMII)
  - 10/100/1000 Mbps Ethernet
  - 802.3af POE
  - used as LAN interface
* 12-24V 1A DC
* internal antennas

Flashing instructions:
======================

Various methods can be used to install the actual image on the flash.
Two easy ones are:

ap51-flash
----------

The tool ap51-flash (https://github.com/ap51-flash/ap51-flash) should be
used to transfer the image to the u-boot when the device boots up.

initramfs from TFTP
-------------------

The serial console must be used to access the u-boot shell during bootup.
It can then be used to first boot up the initramfs image from a TFTP server
(here with the IP 192.168.1.21):

   setenv serverip 192.168.1.21
   setenv ipaddr 192.168.1.1
   tftpboot 0c00000 <filename-of-initramfs-kernel>.bin && bootm $fileaddr

The actual sysupgrade image can then be transferred (on the LAN port) to the
device via

  scp <filename-of-squashfs-sysupgrade>.bin root@192.168.1.1:/tmp/

On the device, the sysupgrade must then be started using

  sysupgrade -n /tmp/<filename-of-squashfs-sysupgrade>.bin

Signed-off-by: Sven Eckelmann <sven@narfation.org>
[rebase, apply shared DTSI/device node, add LED migration]
Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

mpratt14 commented on target/linux/ath79/dts/ar9344_openmesh_mr600.dtsi:
```diff
@@ -0,0 +1,131 @@
+&eth0 {
+	status = "okay";
+
+	pll-data = <0x06000000 0x00000101 0x00001313>;
+
+	mtd-mac-address = <&art 0x0>;
+
+	phy-mode = "rgmii";
+	phy-handle = <&phy0>;
```
see if this works?

```
	pll-data = <0x02000000 0x00000101 0x00001313>;

	mtd-mac-address = <&art 0x0>;

	phy-mode = "rgmii-id";
	phy-handle = <&phy0>;
```

ecsv replied:
Why should I change the pll-data from the original ar711xx mach file? I already have enough problems with the PRs because I've switched the WAN port to the gmac0 (and the mac addresses in this process) for some devices.

mpratt14 replied:
Because there have been vast improvements in the at803x driver since linux 5.1

there is many cases where it doesn't matter what was in ar71xx, it is outdated as a whole

mpratt14 replied:
you can read what I have discovered here

https://github.com/openwrt/openwrt/pull/3505#issuecomment-716050292

ecsv replied:
Ok, then I will change the values in a separate commit but not in the one which tries to re-add the devices.

Btw. thanks for the link. I will try to reference it

mpratt14 replied:
adding a device to ath79 is considered a completely new addition for the most part. there is no value in "preserving" the exact configuration from ar71xx, and then changing it in later commits.

not for me to decide though, you're welcome to ask the developers, but I feel like this is what they would say

if you do run into trouble you can forget about it, since it is not critical, but it should work fine...

ecsv replied:
based on the responses I got from @adschm, it seems like the review is actually involving a comparison with the ar71xx sources. And I've already broken out some newer changes after some discussion in the OM2P PR.

And to be honest, it is rather easy to squash two commits - unsquashing is a lot harder.

mpratt14 replied:
comparison yes, but like Adrian said, as a point of reference, not to make them identical 👍🏽

mpratt14 commented on target/linux/ath79/dts/qca9558_openmesh_mr1750.dtsi:
```diff
@@ -0,0 +1,173 @@
+	};
+};
+
+&eth0 {
+	status = "okay";
+
+	pll-data = <0xae000000 0xa0000101 0xa0001313>;
+
+	mtd-mac-address = <&art 0x0>;
+
+	phy-mode = "rgmii-rxid";
+	phy-handle = <&phy5>;
+
+	gmac-config {
+		device = <&gmac>;
+		rgmii-gmac0 = <1>;
+		rxd-delay = <3>;
+		rxdv-delay = <3>;
+		txd-delay = <0>;
+		txen-delay = <0>;
+	};
+};
```
and this

also, the gmac property has a completely different name for qca955x series you can see here
https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/ag71xx_gmac.c#L77

```
&eth0 {
	status = "okay";

	pll-data = <0x82000000 0x80000101 0x80001313>;

	mtd-mac-address = <&art 0x0>;

	phy-handle = <&phy5>;
	phy-mode = "rgmii-id";

	gmac-config {
		device = <&gmac>;
		rgmii-enabled = <1>;
		rxd-delay = <3>;
		rxdv-delay = <3>;
	};
};
```

ecsv replied:
Thanks, interesting that it still seemed to work. I have to test the devices again.

mpratt14 commented on target/linux/ath79/dts/qca9558_openmesh_mr900.dtsi:
```diff
@@ -0,0 +1,180 @@
+	};
+};
+
+&eth0 {
+	status = "okay";
+
+	pll-data = <0xae000000 0xa0000101 0xa0001313>;
+
+	mtd-mac-address = <&art 0x0>;
+
+	phy-mode = "rgmii-rxid";
+	phy-handle = <&phy5>;
+
+	gmac-config {
+		device = <&gmac>;
+		rgmii-gmac0 = <1>;
+		rxd-delay = <3>;
+		rxdv-delay = <3>;
+		txd-delay = <0>;
+		txen-delay = <0>;
+	};
+};
+
```
same as my other comment

```
&eth0 {
	status = "okay";

	pll-data = <0x82000000 0x80000101 0x80001313>;

	mtd-mac-address = <&art 0x0>;

	phy-handle = <&phy5>;
	phy-mode = "rgmii-id";

	gmac-config {
		device = <&gmac>;
		rgmii-enabled = <1>;
		rxd-delay = <3>;
		rxdv-delay = <3>;
	};
};
```

ecsv replied:
change was added and it still seems to work

mpratt14 commented:
my review is based on my experience with Engenius boards which are made by Senao, OpenMesh is another brand of Senao.

I am assuming all the eth nodes that use `pll-data` are AR803x series switches correct? you should also mention the switch models in the commit messages

ecsv commented:
Please stop saying that OpenMesh is a brand of Senao. OpenMesh was a company which at some point also let Senao (besides other companies) manufacture their hardware.

mpratt14 commented:
Ok sorry I am not trying to be offensive.................

better to say Senao built boards that are rebranded I guess...

adschm commented:
> based on the responses I got from @adschm, it seems like the review is actually involving a comparison with the ar71xx sources. 

Well, yes and no.

I personally see ath79 as a big advantage to get rid of several peculiar things in ar71xx. And yes, in principle, every device is added "like a new device".

Nevertheless, checking with ar71xx makes sense. It helps to spot errors from less careful authors; from my experience, most errors in device support are typically due to lack of care. That's why I always ask on a deviation from ar71xx: sometimes it's an improvement/fix compared to ar71xx, sometimes it's an error in the ath79 proposal. Won't hurt in both cases.

You are actually one of the first people in this situation that I didn't ask to fix ar71xx in parallel to providing a "fixed" ath79 support; previously, with ar71xx still in master and 19.07 relatively fresh, that fate would have hit you.

But in any case, please don't stop to make the ath79 device support "better" than ar71xx where that's possible. ;-)

adschm commented:
This looks a clone of the MR900 family just with different 5G Wifi implementation. I'd prefer to have a shared DTSI for MR900 and this, unless I've overlooked something. (Probably wait until I've merged the other devices.)

ecsv commented:
@adschm 

They are somewhat similar. The only differences in the dtsi are:

```
--- mr900.dtsi	2021-01-19 14:15:59.764459060 +0100
+++ mr1750.dtsi	2021-01-19 14:16:08.028166059 +0100
@@ -40,7 +40,7 @@
 		wifi2g_blue {
 			label = "blue:wifi2g";
 			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
-			linux,default-trigger = "phy0tpt";
+			linux,default-trigger = "phy1tpt";
 		};
 
 		status_green {
@@ -62,7 +62,7 @@
 		wifi5g_blue {
 			label = "blue:wifi5g";
 			gpios = <&gpio 23 GPIO_ACTIVE_LOW>;
-			linux,default-trigger = "phy1tpt";
+			linux,default-trigger = "phy0tpt";
 		};
 	};
 
@@ -168,11 +168,4 @@
 
 &pcie0 {
 	status = "okay";
-
-	wifi@0,0 {
-		compatible = "pci168c,0033";
-		reg = <0x0000 0 0 0 0>;
-		mtd-mac-address = <&art 0x0>;
-		mtd-mac-address-increment = <16>;
-	};
 };
```

So nothing which cannot be /delete-node/'ed or just overwritten in the specific dts files.

adschm committed:
ath79: consolidate common definitions for OpenMesh devices

The shared image definitions for OpenMesh devices are currently
organized based on device families. This introduces some duplicate
code, as the image creation code is mostly the same for those.

This patch thus derives two basic shared definitions that work for
all devices and only requires a few variables to be moved back to
the device definitions.

Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

adschm committed:
ath79: make OpenMesh MR900 DTSI more general

The OpenMesh MR900 and to-be-added MR1750 family are very similar.
Make the existing MR900 DTSI more general so it can be used for
the MR1750 devices as well.

Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

mpratt14 commented:
Simply put the `leds` and `&pcie0` nodes only in both DTS and not in the DTSI 👍🏽

ecsv commented:
@mpratt14 You mean in all 4? Not a big fan of that suggestion

mpratt14 commented:
hmmm.... well that is the easier way to implement but it would not save as many lines

I guess it depends on how many leds are similar

DTSI can have `leds` node with only the common ones, and DTS with `&leds` node with the non-common ones

adschm commented:
I've planned another adjustment as well and will modify the files here myself and provide them for review when I finished my build-testing....

adschm commented:
Okay, master is updated.

I've made an attempt to generalize the common device definitions, and added the updated on top of that in my branch here:

https://github.com/adschm/openwrt/commits/ombuild

I'd be happy if you could review/test this, and if you are happy with it I'd merge it without you needing to update this PR anymore ...

Obviously, it might save some time to settle this before updating the other PRs ...

ecsv commented:
https://github.com/adschm/openwrt/commit/340daa00575836b91c9d05ea58a2ab2eb448fedd seems to be what I had in mind too.

I have to check the rest (especially "ath79: consolidate common definitions for OpenMesh devices") in detail when I am home again

ecsv commented:
@adschm: Changes look fine to me.

adschm commented:
> @adschm: Changes look fine to me.

So I can push my version to master? Just wondering because you pushed them here ...

ecsv commented:
@adschm They are the same changes. So just push your branch.

Just pushed them here to keep track of it while adjusting the OM2P, OM5P, A60, OM5PAC PRs. I have to shuffle a lot around because some of the devices (OM2Pv1, OM5P-AN) were skipped.

adschm commented:
> @adschm They are the same changes. So just push your branch.

Okay, thanks, will do that now.

> Just pushed them here to keep track of it while adjusting the OM2P, OM5P, A60, OM5PAC PRs. I have to shuffle a lot around because some of the devices (OM2Pv1, OM5P-AN) were skipped.

Submitting so many devices at the same time is never easy. You can just choose _how_ you have to adapt, not _if_ ... ;-)
